File: py/Day02.py
import re
import sys
import time


def check_line(line: str) -> bool:
    splits = line.split(" ")
    increasing = None
    
    for i in range(1, len(splits)):
    
        current = int(splits[i - 1])
        next_val = int(splits[i])
        
        if increasing is None and next_val != current:
            increasing = next_val > current
        
        if current == next_val:
            break
        
        if increasing:
            if next_val - current <= 0:
                break
            if next_val - current > 3:
                break
        else:
            if current - next_val <= 0:
                break
            if current - next_val > 3:
                break
            
        if i+1 == len(splits):
            i += 1
    
    return i == len(splits)


def main():
    time_start = int(time.time() * 1000)
    file_path = r'..\input_2024\day02_data.txt'
    
    result_part_1 = 0
    
    with open(file_path, 'r') as file:
        for line in file:
            if check_line(line):
                result_part_1 += 1
    
    print("result part 1:", result_part_1)
    
    
    result_part_2 = 0    
    with open(file_path, 'r') as file:
        for line in file:
            if check_line(line):
                result_part_2 += 1
            else:
                splits = line.split(" ")
                
                for j in range(0, len(splits)):
                    new_line="";
                    for k in range(0, len(splits)):
                        if j!=k:
                            new_line += splits[k]
                            new_line += " "
                    
                    if check_line(new_line.strip()):
                        result_part_2 += 1
                        break;
    
    
    print("result part 2:", result_part_2)
    
    time_end = int(time.time() * 1000)
    print(f"Execution time: {time_end - time_start} msec")
    

if __name__ == "__main__":
    main()
    
    

File: src/AoC2024/Day02.java
package AoC2024;

import AoC2022.Day;

public class Day02 extends Day {

	public Day02() {
		super();
	}
	
	public boolean checkLine(String line) {
		String[] splits = line.split(" ");
		
		long current;
		long next;
		
		Boolean increasing = null;
		
		int i = 1;
		for (i = 1; i < splits.length; i++) {
			current = Long.parseLong(splits[i-1]);
			next = Long.parseLong(splits[i]);
			
			if(increasing==null) increasing = Boolean.valueOf(next>current);
			if(current==next) break;
			
			if(increasing) {
				if(next-current<=0) break;
				if(next-current>3) break;
			} else {
				if(current-next<=0) break;
				if(current-next>3) break;
			}
		}
		return i==splits.length;
	}
	
	public void processFile() {
		String[] lines = sb.toString().split("\n");
		long result_part1 = 0;
		for (String line : lines) {
			
			if(checkLine(line)) result_part1++; // corect: 606
			
		}
		
		logln("result part1: "+result_part1); // correct: 202
		
		
		lines = sb.toString().split("\n");
		long result_part2 = 0;
		for (String line : lines) {
			
			if(checkLine(line)) {
				result_part2++;
			} else {
				String[] splits = line.split(" ");
				
				String newLine="";
				for (int i = 0; i < splits.length; i++) {
					newLine="";
					for (int j = 0; j < splits.length; j++) {
						if(i!=j) {
							newLine += splits[j];
							newLine += " ";
						}
					}
					
					if(checkLine(newLine)) {
						result_part2++;
						break;
					}
					
				}
			}
			
		}
		
		logln("result part2: "+result_part2); // correct: 644
		
	}

	public static void main(String[] args) {
		long time_start = System.currentTimeMillis();
		Day02 day = new Day02();
		day.readFile();
		day.processFile();
		long time_end = System.currentTimeMillis();
		day.logln("Execution time: "+(time_end-time_start)+" msec");
		day.flush();
	}

}
